src/util.ts

```diff
@@ -35,3 +37,42 @@ export function normalize<T = {}, U = Function>(
 export function objectEntries<T>(obj: {[key: string]: T}): Array<[string, T]> {
   return Object.keys(obj).map(key => [key, obj[key]] as [string, T]);
 }
+
+/**
+ * URI encode the given string for generating signed URLs:
+ * Encode every byte except `A-Z a-Z 0-9 ~ - . _`.
+ *
+ * encodeURI patches encodeURIComponent() by:
+ *  - additionally encoding `! * ' ( )` characters;
+ *  - conditionally encoding `/` if encodeSlash is `true`.
+ * @param {string} uri The URI to encode.
+ * @param [boolean=false] encodeSlash If `true`, the "/" character is not encoded.
+ * @return {string} The encoded string.
+ */
+export function encodeURI(uri: string, encodeSlash: boolean): string {
```

Comment:
I find this to be a little bit weird, can we do a little bit of research and figure out what library the upstream API is using for encoding, I'd be interested in finding out if the deltas between the upstream API's URI encoding, and `encodeURIComponent`'s encoding are documented anywhere, e.g., is this a difference in how the Node.js library and a Java library encode URIs, is there a specification that either is based on?

Reply:
Implementations of built-in URI encoding vary between languages and the service. In this particular case, the encoding expected by the GCS service is to encode everything except `A-Z a-Z 0-9 ~ - . _` while `encodeURIComponent()` excludes these plus `! ' ( ) *` (the delta). This is basically `fixedEncodeURIComponent()` in [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) MDN doc.

Reply:
I'm convinced :+1: sounds like something that could potentially be addressed at the language level eventually; perhaps add a comment pointing to the MDN doc? (It's useful context that`encodeURIComponent` isn't spec compliant).